File: engines/jena/src/main/kotlin/Engine.kt
package es.weso.rudof

import org.apache.jena.geosparql.configuration.GeoSPARQLConfig
import org.apache.jena.graph.Graph
import org.apache.jena.riot.Lang
import org.apache.jena.riot.RDFDataMgr
import org.apache.jena.shacl.Shapes
import org.apache.jena.shacl.ValidationReport
import org.apache.jena.shacl.validation.ShaclPlainValidator
import java.io.StringWriter

class Engine : ValidationEngine<ValidationReport> {
    override val name = "jena"

    private lateinit var shapes: Shapes
    private lateinit var dataGraph: Graph
    private var sizeLogged = false

    init {
        GeoSPARQLConfig.setupMemoryIndex()
    }

    override fun loadData(dataPath: String, dataFormat: String, shapesPath: String, shapesFormat: String) {
        dataGraph = RDFDataMgr.loadGraph("file:$dataPath")
        val shapesGraph = RDFDataMgr.loadGraph("file:$shapesPath")

        if (!sizeLogged) {
            println("[$name] Data graph size:   ${dataGraph.size()}")
            println("[$name] Shapes graph size: ${shapesGraph.size()}")
            sizeLogged = true
        }

        shapes = Shapes.parse(shapesGraph)
    }

    override fun validate(): ValidationReport =
        ShaclPlainValidator().validate(shapes, dataGraph)

    override fun generateReport(result: ValidationReport): String =
        StringWriter().use { os ->
            RDFDataMgr.write(os, result.model, Lang.TURTLE)
            os.toString()
        }
}

File: engines/pyshacl/engine.py
import rdflib
from pyshacl import validate
from rdflib import Graph

from validation_engine import ValidationEngine

class Engine(ValidationEngine[Graph]):
    name = "pyshacl"

    def __init__(self) -> None:
        self._data_graph: Graph | None = None
        self._shapes_graph: Graph | None = None
        self._size_logged = False

    def load_data(self, data_path: str, data_format: str, shapes_path: str, shapes_format: str) -> None:
        self._data_graph = rdflib.Graph()
        assert self._data_graph is not None
        with open(data_path, "r", encoding="utf-8") as f:
            self._data_graph.parse(data=f.read())

        self._shapes_graph = rdflib.Graph()
        assert self._shapes_graph is not None
        with open(shapes_path, "r", encoding="utf-8") as f:
            self._shapes_graph.parse(data=f.read())

        if not self._size_logged:
            print(f"[{self.name}] Data graph size:   {len(self._data_graph)}")
            print(f"[{self.name}] Shapes graph size: {len(self._shapes_graph)}")
            self._size_logged = True

    def validate(self) -> Graph:
        assert self._data_graph is not None
        assert self._shapes_graph is not None
        _, results_graph, _ = validate(data_graph=self._data_graph, shacl_graph=self._shapes_graph, inference="")
        return results_graph

    def generate_report(self, result: Graph) -> str:
        return result.serialize(format="turtle")
